README.md
# rgwadmin

[![Documentation Status](https://readthedocs.org/projects/rgwadmin/badge/?version=latest)](https://rgwadmin.readthedocs.io/en/latest/?badge=latest)

rgwadmin is a Python library to access the Ceph Object Storage Admin API.

http://docs.ceph.com/docs/master/radosgw/adminops/


## API Example Usage

```python
from rgwadmin import RGWAdmin

rgw = RGWAdmin(access_key='XXX', secret_key='XXX', server='obj.example.com')
rgw.create_user(
    uid='liam',
    display_name='Liam Monahan',
    email='liam@umiacs.umd.edu',
    user_caps='usage=read, write; users=read',
    max_buckets=1000)
rgw.set_user_quota(
    uid='liam',
    quota_type='user',
    max_size_kb=1024*1024,
    enabled=True)
rgw.remove_user(uid='liam', purge_data=True)
```

## User Example Usage
```python
from rgwadmin import RGWAdmin, RGWUser

RGWAdmin.connect(access_key='XXX', secret_key='XXX', server='obj.example.com')
u = RGWUser.create(user_id='test', display_name='Test User')
u.user_quota.size = 1024 * 1024  # in bytes
u.user_quota.enabled = True
u.save()
u.delete()
```

## Requirements

rgwadmin requires the following Python packages:

 * [requests](http://python-requests.org/)
 * [requests-aws](https://github.com/tax/python-requests-aws)

 If you are running python < 2.7 you will need a backport of collections
 ordereddict.
 * [ordereddict](https://pypi.python.org/pypi/ordereddict)

Additionally, you need to have a [Ceph](http://www.ceph.org) Object Storage
instance with a user that has appropriate caps (capabilities) on the parts of
the API that you want to access.  See the
[Ceph Object Storage](http://docs.ceph.com/docs/master/radosgw/) page for more
information.

### Compatibility
The API for different versions of Ceph RadosGW has changed over time.  While we
endeavor to be compatible with as many versions of Ceph RadosGW there will be
some issues that crop up over time.  We only consider the LTS versions but
these are what we have tested on.

* 1.0.5 <= Firefly
* 1.0.6 > Firefly/Hammer
* 1.1 > Hammer
* 1.1.5 > Jewel

## Installation

```pip install rgwadmin```


## License

    rgwadmin - a Python interface to the Rados Gateway Admin API
    Copyright (C) 2015  UMIACS

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

    Email:
        github@umiacs.umd.edu


CHANGELOG.md
# Changelog

Please refer to https://github.com/UMIACS/rgwadmin/releases instead.


test.py
#!/usr/bin/env python

import os
import logging
import rgwadmin
import unittest
import random
from rgwadmin.exceptions import BucketAlreadyExists

logging.basicConfig(level=logging.WARNING)

access_key = os.environ['OBJ_ACCESS_KEY_ID']
secret_key = os.environ['OBJ_SECRET_ACCESS_KEY']
server = os.environ['OBJ_SERVER']


class RGWAdminTest(unittest.TestCase):

    def setUp(self):
        self.rgw = rgwadmin.RGWAdmin(access_key,
                                     secret_key,
                                     server,
                                     secure=True)
        self.user1 = 'foo1209'
        self.user2 = 'foo1213'
        self.user3 = 'bar3142'
        self.secret = rgwadmin.RGWAdmin.gen_secret_key()
        user1 = self.rgw.create_user(uid=self.user1,
                                     email='%s@example.com' % self.user1,
                                     display_name='Unit Test %s' % self.user1,
                                     secret_key=self.secret)
        user2 = self.rgw.create_user(uid=self.user2,
                                     display_name='Unit Test %s' % self.user2,
                                     secret_key=self.secret)
        self.user1_obj = user1
        self.user2_obj = user2
        self.assertTrue(user1['user_id'] == self.user1)
        self.assertTrue(user2['user_id'] == self.user2)

    def tearDown(self):
        self.rgw.remove_user(uid=self.user1)
        self.rgw.remove_user(uid=self.user2)

    def test_modify_user(self):
        user = self.rgw.modify_user(uid=self.user1,
                                    email='%s@test.com' % self.user1)
        self.assertTrue(user['email'] == '%s@test.com' % self.user1)

    def test_duplicate_email(self):
        # NOTE: this exception is wrong and should be fixed at some point
        # in the future in Ceph http://tracker.ceph.com/issues/13635
        with self.assertRaises(BucketAlreadyExists):
            self.rgw.create_user(uid=self.user3,
                                 email='%s@example.com' % self.user1,
                                 display_name='Unit Test %s' % self.user3,
                                 secret_key=rgwadmin.RGWAdmin.gen_secret_key())

    def test_get_user(self):
        user = self.rgw.get_user(uid=self.user2)
        self.assertTrue(user['display_name'] == 'Unit Test %s' % self.user2)

    def test_get_users(self):
        users = self.rgw.get_users()
        self.assertTrue(self.user1 in users)
        self.assertTrue(self.user2 in users)

    def test_user_quota(self):
        size = random.randint(1000, 1000000)
        self.rgw.set_quota(uid=self.user1, quota_type='user',
                           max_size_kb=size, enabled=True)
        user1_quota_info = self.rgw.get_user_quota(uid=self.user1)
        self.assertTrue(size == user1_quota_info['max_size_kb'])

    def test_bucket(self):
        bucket_name = self.user1 + '_bucket'
        self.rgw.create_bucket(bucket=bucket_name)
        bucket = self.rgw.get_bucket(bucket=bucket_name)
        self.rgw.link_bucket(bucket=bucket_name, bucket_id=bucket['id'],
                             uid=self.user1)
        self.rgw.get_bucket(uid=self.user1, bucket=bucket_name)
        self.rgw.get_policy(bucket=bucket_name)
        self.rgw.remove_bucket(bucket=bucket_name, purge_objects=True)
        self.assertTrue(True)

    def test_get_usage(self):
        summary = self.rgw.get_usage(show_summary=True)
        self.assertTrue('summary' in summary)

    def test_subuser(self):
        self.rgw.create_subuser(uid=self.user2,
                                subuser='swift',
                                key_type='swift',
                                secret_key=self.secret)
        self.rgw.modify_subuser(uid=self.user2,
                                subuser='swift',
                                access='write')
        subuser = self.rgw.get_user(uid=self.user2)
        self.rgw.remove_subuser(uid=self.user2, subuser='swift')
        for subs in subuser['subusers']:
            if subs['id'] == '%s:%s' % (self.user2, 'swift'):
                self.assertTrue(subs['permissions'] == 'write')

    def test_s3_keys(self):
        access = rgwadmin.RGWAdmin.gen_secret_key(size=20)
        secret = rgwadmin.RGWAdmin.gen_secret_key(size=40)
        keys = self.rgw.create_key(uid=self.user1,
                                   access_key=access,
                                   secret_key=secret)
        for key in keys:
            if key['access_key'] == access:
                self.assertTrue(key['secret_key'] == secret)


if __name__ == '__main__':
    unittest.main()


test/__init__.py
def create_bucket(rgw, bucket=None):
    """
    This is a helper function for tests to create buckets so they can assert changes
    on the gateway based on the bucket's existence.

    NOTE: The PUT /:bucket endpoint is not part of the Admin Ops API, and therefore
    is not included in the rgw class.  The response format does not match other
    Admin Operations, and it is largely incompatible with the rgw class.
    """
    return rgw.request('put', '/%s' % bucket)


test/test_rgw.py
#!/usr/bin/env python

import logging
import rgwadmin
import unittest
import random
from rgwadmin.exceptions import InvalidArgument
from rgwadmin.utils import get_environment_creds
from . import create_bucket

logging.basicConfig(level=logging.WARNING)


class RGWAdminTest(unittest.TestCase):

    def setUp(self):
        self.rgw = rgwadmin.RGWAdmin(
            secure=False, verify=False, **get_environment_creds())
        self.user1 = 'foo1209'
        self.user2 = 'foo1213'
        self.user3 = 'bar3142'
        self.secret = rgwadmin.RGWAdmin.gen_secret_key()
        user1 = self.rgw.create_user(uid=self.user1,
                                     email='%s@example.com' % self.user1,
                                     display_name='Unit Test %s' % self.user1,
                                     secret_key=self.secret)
        user2 = self.rgw.create_user(uid=self.user2,
                                     display_name='Unit Test %s' % self.user2,
                                     secret_key=self.secret)
        self.user1_obj = user1
        self.user2_obj = user2
        self.assertTrue(user1['user_id'] == self.user1)
        self.assertTrue(user2['user_id'] == self.user2)

    def tearDown(self):
        self.rgw.remove_user(uid=self.user1)
        self.rgw.remove_user(uid=self.user2)

    def test_modify_user(self):
        before_keys = self.user1_obj['keys']
        user = self.rgw.modify_user(uid=self.user1,
                                    email='%s@test.com' % self.user1)
        self.assertTrue(user['email'] == '%s@test.com' % self.user1)

        # ensure that generate_key is False and that new keys are not
        # being created by default when modifying the user.
        self.assertEqual(before_keys, user['keys'])

        # generate_key=True should add an extra pair of keys
        user = self.rgw.modify_user(uid=self.user1, generate_key=True)
        self.assertNotEqual(before_keys, user['keys'])
        self.assertEqual(len(before_keys) + 1, len(user['keys']))

        # modifying a user doesn't automatically un-suspend them
        # setup - suspend them first
        user = self.rgw.modify_user(uid=self.user1, suspended=True)
        self.assertEqual(user['suspended'], 1)

        # test - modify their username without unsuspending them
        user = self.rgw.modify_user(uid=self.user1,
                                    email='%s@test2.com' % self.user1)
        self.assertTrue(user['email'] == '%s@test2.com' % self.user1)
        self.assertEqual(user['suspended'], 1)

        # unsuspend the user to allow other tests to function
        user = self.rgw.modify_user(uid=self.user1, suspended=False)
        self.assertEqual(user['suspended'], 0)

    def test_duplicate_email(self):
        with self.assertRaises(InvalidArgument):
            self.rgw.create_user(uid=self.user3,
                                 email='%s@example.com' % self.user1,
                                 display_name='Unit Test %s' % self.user3,
                                 secret_key=rgwadmin.RGWAdmin.gen_secret_key())

    def test_get_user(self):
        user = self.rgw.get_user(uid=self.user2)
        self.assertTrue(user['display_name'] == 'Unit Test %s' % self.user2)

    def test_get_users(self):
        users = self.rgw.get_users()
        self.assertTrue(self.user1 in users)
        self.assertTrue(self.user2 in users)

    def test_user_quota(self):
        size = random.randint(1000, 1000000)
        self.rgw.set_user_quota(uid=self.user1, quota_type='user',
                                max_size_kb=size, enabled=True)
        user1_quota_info = self.rgw.get_user_quota(uid=self.user1)
        self.assertTrue(size == user1_quota_info['max_size_kb'])

    def test_bucket_quota(self):
        size = random.randint(1000, 1000000)
        bucket_name = self.user1 + '_bucket'
        create_bucket(self.rgw, bucket=bucket_name)
        self.rgw.set_bucket_quota(uid=self.user1, bucket=bucket_name,
                                  max_size_kb=size, enabled=True)
        bucket = self.rgw.get_bucket(bucket=bucket_name)
        self.assertTrue(bucket['bucket_quota']['max_size_kb'] == size)

    def test_bucket(self):
        bucket_name = self.user1 + '_bucket'
        create_bucket(self.rgw, bucket=bucket_name)
        bucket = self.rgw.get_bucket(bucket=bucket_name)
        self.rgw.link_bucket(bucket=bucket_name, bucket_id=bucket['id'],
                             uid=self.user1)
        self.rgw.get_bucket(uid=self.user1, bucket=bucket_name)
        self.rgw.get_policy(bucket=bucket_name)
        self.rgw.remove_bucket(bucket=bucket_name, purge_objects=True)
        self.assertTrue(True)

    def test_get_usage(self):
        summary = self.rgw.get_usage(show_summary=True)
        self.assertTrue('summary' in summary)

    def test_subuser(self):
        self.rgw.create_subuser(uid=self.user2,
                                subuser='swift',
                                key_type='swift',
                                secret_key=self.secret)
        self.rgw.modify_subuser(uid=self.user2,
                                subuser='swift',
                                access='write')
        subuser = self.rgw.get_user(uid=self.user2)
        self.rgw.remove_subuser(uid=self.user2, subuser='swift')
        for subs in subuser['subusers']:
            if subs['id'] == '%s:%s' % (self.user2, 'swift'):
                self.assertTrue(subs['permissions'] == 'write')

    def test_s3_keys(self):
        access = rgwadmin.RGWAdmin.gen_secret_key(size=20)
        secret = rgwadmin.RGWAdmin.gen_secret_key(size=40)
        keys = self.rgw.create_key(uid=self.user1,
                                   access_key=access,
                                   secret_key=secret)
        for key in keys:
            if key['access_key'] == access:
                self.assertTrue(key['secret_key'] == secret)

    def test_parse_rados_datestring(self):
        rgwadmin.RGWAdmin.parse_rados_datestring(u'2016-06-27T16:06:39.163Z')


if __name__ == '__main__':
    unittest.main()


rgwadmin/utils.py
import os
import string
import random


def random_password(size=40, chars=None):
    '''Return a random password string.

    The default length is 32 characters.  Different character classes can be
    passed in, but the default draws from ASCII uppercase, lowercase, and
    digits.

    Uses the SystemRamdom to ensure you get a trustworty entropy pool.
    '''
    if chars is None:
        chars = string.ascii_uppercase + string.ascii_lowercase + string.digits
    random_chars = random.SystemRandom().choice
    return ''.join(random_chars(chars) for _ in range(size))


def get_environment_creds():
    secure = os.environ.get('OBJ_SERVER_SECURE', True)
    if secure is not True and secure == 'false':
        secure = False
    verify = os.environ.get('OBJ_SERVER_VERIFY', True)
    if verify is not True and verify == 'false':
        verify = False
    return {'access_key': os.environ['OBJ_ACCESS_KEY_ID'],
            'secret_key': os.environ['OBJ_SECRET_ACCESS_KEY'],
            'server': os.environ['OBJ_SERVER'],
            'port': os.environ.get('OBJ_SERVER_PORT', 443),
            'secure': secure,
            'verify': verify}


def id_generator(size=6, chars=None):
    if chars is None:
        chars = string.ascii_lowercase + string.ascii_uppercase + string.digits
    letter = random.choice(string.ascii_lowercase)
    return letter + ''.join(random.choice(chars) for x in range(size))


rgwadmin/exceptions.py
"""
rgwadmin.exceptions
~~~~~~~~~~~~~~~~~~~

This module contains the rgwadmin exceptions.

"""


class RGWAdminException(Exception):
    """There was a unlabeled exception that was raised during your request"""
    def __init__(self, code, raw=None):
        self.code = code
        self.raw = raw


class AccessDenied(RGWAdminException):
    """Access was denied for the request."""


class UserExists(RGWAdminException):
    """Attempt to create existing user."""


class InvalidAccessKey(RGWAdminException):
    """Invalid access key specified."""


class InvalidArgument(RGWAdminException):
    """Invalid argument specified."""


class InvalidKeyType(RGWAdminException):
    """Invalid key type specified."""


class InvalidSecretKey(RGWAdminException):
    """Invalid secret key specified."""


class KeyExists(RGWAdminException):
    """Provided access key exists and belongs to another user."""


class EmailExists(RGWAdminException):
    """Provided email address exists."""


class SubuserExists(RGWAdminException):
    """Specified subuser exists."""


class InvalidAccess(RGWAdminException):
    """Invalid subuser access specified."""


class IndexRepairFailed(RGWAdminException):
    """Bucket index repair failed."""


class BucketNotEmpty(RGWAdminException):
    """Attempted to delete non-empty bucket."""


class ObjectRemovalFailed(RGWAdminException):
    """Unable to remove objects."""


class BucketUnlinkFailed(RGWAdminException):
    """Unable to unlink bucket from specified user."""


class BucketLinkFailed(RGWAdminException):
    """Unable to link bucket to specified user."""


class NoSuchObject(RGWAdminException):
    """Specified object does not exist."""


class IncompleteBody(RGWAdminException):
    """Either bucket was not specified for a bucket policy request or bucket
       and object were not specified for an object policy request."""


class InvalidCap(RGWAdminException):
    """Attempt to grant invalid admin capability."""


class NoSuchCap(RGWAdminException):
    """User does not possess specified capability."""


class InternalError(RGWAdminException):
    """Internal server error."""


class NoSuchUser(RGWAdminException):
    """User does not exist."""


class NoSuchBucket(RGWAdminException):
    """Bucket does not exist."""


class NoSuchKey(RGWAdminException):
    """No such access key."""


class ServerDown(RGWAdminException):
    """The backing server is not available."""


class InvalidQuotaType(RGWAdminException):
    """You must specify either a 'user' or 'bucket' quota type"""


class BucketAlreadyExists(RGWAdminException):
    """The bucket already exists"""


rgwadmin/user.py
import logging
try:
    from collections import OrderedDict
except ImportError:
    from ordereddict import OrderedDict
from .utils import random_password
from .rgw import RGWAdmin
from .exceptions import NoSuchKey

log = logging.getLogger(__name__)

try:
    unicode
except NameError:
    unicode = str


class AttributeMixin(object):
    attrs = []

    def __str__(self):
        try:
            from qav.listpack import ListPack
        except:
            return str(self.to_tuples())
        return str(ListPack(self.to_tuples()))

    def to_tuples(self):
        return [(x, getattr(self, x)) for x in self.attrs]

    def to_dict(self):
        return dict(self.to_tuples())


class RGWCap(AttributeMixin):
    attrs = ['type', 'perm']

    def __init__(self, type, perm):
        self.type = type
        self.perm = perm

    def __repr__(self):
        return str('%s %s' % (self.__class__.__name__, self.type))


class RGWKey(AttributeMixin):
    attrs = ['user', 'access_key', 'secret_key']

    def __init__(self, user, access_key, secret_key):
        self.user = user
        self.access_key = access_key
        self.secret_key = secret_key

    def __repr__(self):
        return str('%s %s' % (self.__class__.__name__, self.user))

    @staticmethod
    def generate(user, access_size=20, secret_size=40):
        access_key = random_password(size=access_size)
        secret_key = random_password(size=secret_size)
        return RGWKey(user, access_key, secret_key)


class RGWSubuser(AttributeMixin):
    attrs = ['id', 'permissions']
    permission_list = ['read', 'write', 'read-write', 'full-control']

    def __init__(self, id, permissions):
        self.id = id
        self.permissions = permissions

    def __repr__(self):
        return str('%s %s' % (self.__class__.__name__, self.id))

    @property
    def permissions(self):
        return self._permissions

    @permissions.setter
    def permissions(self, value):
        if value in self.permission_list:
            self._permissions = value
        elif value == '<none>':
            self._permissions = None
        else:
            raise AttributeError


class RGWSwiftKey(AttributeMixin):
    attrs = ['user', 'secret_key']

    def __init__(self, user, secret_key):
        self.user = user
        self.secret_key = secret_key

    def __repr__(self):
        return str('%s %s' % (self.__class__.__name__, self.user))

    @staticmethod
    def generate(user, secret_size=40):
        secret_key = random_password(size=secret_size)
        return RGWSwiftKey(user, secret_key)


class RGWQuota(AttributeMixin):
    attrs = ['enabled', 'max_objects', 'max_size_kb']
    defaults = OrderedDict([('enabled', False),
                            ('max_objects', -1),
                            ('max_size_kb', -1)])

    def __init__(self, **kwargs):
        for attr in self.attrs:
            setattr(self, attr, kwargs[attr])

    def __repr__(self):
        return '%s %s/%s' % (self.enabled, self.max_objects,
                             self.string_size())

    def string_size(self):
        weights = ['KB', 'MB', 'GB', 'TB', 'PB']
        size = int(self.max_size_kb)
        weight = 0
        if size < 0:
            return size
        while True:
            if size < 1024 or weight == len(weights):
                return str('%s%s' % (size, weights[weight]))
            else:
                size = size / 1024
                weight += 1

    @classmethod
    def default(cls):
        return cls(**cls.defaults)

    @property
    def size(self):
        return int(self.max_size_kb) * 1000

    @size.setter
    def size(self, new_size):
        self.max_size_kb = int(new_size) / 1000


class RGWUser(AttributeMixin):
    attrs = ['user_id', 'display_name', 'email', 'caps', 'keys',
             'max_buckets', 'suspended', 'swift_keys', 'subusers',
             'placement_tags', 'bucket_quota', 'user_quota',
             'default_placement', 'op_mask', 'temp_url_keys']
    modify_attrs_mask = ['placement_tags', 'default_placement',
                         'op_mask', 'temp_url_keys']
    sub_attrs = OrderedDict([('caps', RGWCap),
                             ('keys', RGWKey),
                             ('swift_keys', RGWSwiftKey),
                             ('subusers', RGWSubuser),
                             ('bucket_quota', RGWQuota),
                             ('user_quota', RGWQuota)])

    """Representation of a RadosGW User"""
    def __init__(self, **kwargs):
        self.sensitive_attrs = [('keys', 'secret_key')]
        for attr in self.attrs:
            setattr(self, attr, kwargs[attr])
        if kwargs.get('sensitive_attrs'):
            self.sensitive_attrs.append(kwargs.get('sensitive_attrs'))

    def __repr__(self):
        return str('%s %s' % (self.__class__.__name__, self.user_id))

    @classmethod
    def create(cls, user_id, display_name, **kwargs):
        rgw = RGWAdmin.get_connection()
        log.debug('Creating user %s' % user_id)
        rgw.create_user(uid=user_id,
                        display_name=display_name,
                        **kwargs)
        return cls.fetch(user_id)

    def diff(self):
        rgw = RGWAdmin.get_connection()
        d = {}
        current = self.to_dict()
        try:
            existing = rgw.get_metadata('user', self.user_id)['data']
        except NoSuchKey:
            logging.info('Unable to fetch the user %s.' % self.user_id)
            return current
        for k in current.keys():
            if k in existing.keys():
                if current[k] != existing[k]:
                    d[k] = (existing[k], current[k])
        return d

    def exists(self):
        """Return True if the user exists.  False otherwise."""
        rgw = RGWAdmin.get_connection()
        try:
            rgw.get_metadata('user', self.user_id)
            return True
        except NoSuchKey:
            return False

    def save(self):
        rgw = RGWAdmin.get_connection()

        if not self.exists():
            log.debug('User does not exist. Creating %s' % self.user_id)
            rgw.create_user(uid=self.user_id,
                            display_name=self.display_name)

        d = self._modify_dict()
        log.debug('Modify existing user %s %s' % (self.user_id, d))
        rgw.modify_user(**d)
        rgw.set_user_quota(self.user_id, 'user', **self.user_quota.to_dict())
        rgw.set_user_quota(self.user_id, 'bucket', **self.bucket_quota.to_dict())

    def delete(self):
        rgw = RGWAdmin.get_connection()
        try:
            rgw.get_metadata('user', self.user_id)
        except NoSuchKey:
            logging.error('User %s does not exist yet aborting.' %
                          self.user_id)
            return False
        logging.info('Deleting user %s from %s' %
                     (self.user_id, rgw.get_base_url()))
        return rgw.remove_user(uid=self.user_id)

    @classmethod
    def list(cls):
        rgw = RGWAdmin.get_connection()
        return [cls.fetch(x) for x in rgw.get_users()]

    @classmethod
    def fetch(cls, user):
        log.debug('Fetching user %s' % user)
        rgw = RGWAdmin.get_connection()
        try:
            j = rgw.get_metadata('user', user)
        except NoSuchKey:
            return None
        else:
            return cls._parse_user(j['data'])

    @classmethod
    def _parse_user(cls, rgw_user):
        # we expect to be passed a dict
        if type(rgw_user) is not dict:
            log.warning('rgw_user is not a dict instance')
            return None
        # check to make sure we have all the correct keys
        if not set(map(lambda x: unicode(x), cls.attrs)) <= \
                set(rgw_user.keys()):
            return None
        log.debug('Parsing RGWUser %s' % rgw_user)
        for subattr in cls.sub_attrs.keys():
            log.debug('Loading attribute %s with class %s' %
                      (subattr, cls.sub_attrs[subattr].__name__))
            if type(rgw_user[subattr]) is list:
                obj = [cls.sub_attrs[subattr](**x) for x in rgw_user[subattr]]
            elif type(rgw_user[subattr]) is dict:
                obj = cls.sub_attrs[subattr](**rgw_user[subattr])
            else:
                obj = rgw_user[subattr]
            rgw_user[subattr] = obj
        return RGWUser(**rgw_user)

    def _scrubbed_dict(self):
        '''Return a dict representation of the object with sensitve attrs
           filtered.
        '''
        scrubbed = self.to_dict()
        censor = '******'
        for k, v in self.sensitive_attrs:
            if type(scrubbed[k]) is list:
                for o in scrubbed[k]:
                    o[v] = censor
            else:
                scrubbed[k] = censor
        return scrubbed

    def _modify_dict(self):
        '''Return flat dict representation of the object'''
        d = {}
        for attr in self.attrs:
            if attr not in self.modify_attrs_mask+list(self.sub_attrs.keys()):
                d[attr] = getattr(self, attr)
        d['uid'] = d.pop('user_id')
        d['generate_key'] = False
        return d

    def to_dict(self):
        '''Return the dict representation of the object'''
        d = {}
        for attr in self.attrs:
            if attr in self.sub_attrs.keys():
                if type(getattr(self, attr)) is list:
                    d[attr] = []
                    for o in getattr(self, attr):
                        d[attr].append(o.to_dict())
                else:
                    d[attr] = getattr(self, attr).to_dict()
            else:
                d[attr] = getattr(self, attr)
        return d
